### PR TITLE
feat(temperature): Add `zone-type` setting

### PR DESCRIPTION
**Pull request by xphoniex**
## What type of PR is this? (check all applicable)

* [x] Feature
* [x] Bug Fix

## Description

CPU's `thermal-zone` changes on every boot on my machine, sometimes it's `1` and other times it's `2`. Others also have had this issue and resorted to scripts updating the `thermal-zone` before starting `polybar`.

This can simply be fixed using a `zone-type` config option:

If you look at the content in `/sys/class/thermal/thermal_zone%zone%/type`, you'll get the name of the sensor and you can easily deduct zone from that. For example for Intel CPUs, you should set `zone-type = x86_pkg_temp` and you'll be set.

## Related Issues & Documents

Ref #2572

## Documentation (check all applicable)

* [x] This PR requires changes to the Wiki documentation (describe the changes)

### CODE REVIEW

**Comment by patrick96** [comment]
I'm not around right now, so can't really review this. Will review once I'm back and have time. But general idea looks good. Except that you just try the first 10 zones instead of listing files in the folder and iterating through those.

**Comment by xphoniex** [comment]
> I'm not around right now, so can't really review this. Will review once I'm back and have time. But general idea looks good. Except that you just try the first 10 zones instead of listing files in the folder and iterating through those.

Okay, cool.

Re iterations, rewrote it to use `INT_MAX` ([here](https://github.com/polybar/polybar/compare/eb369e41f99b73a63f0a85d28277d05c91ed79d5..5b96d0f01dc08d9d434b6fc5baaa76e0d472b92e)), this is simpler than listing the files (no allocation needed, and no need to then extract zone # from the file string).

I have a few useful contributions which I can open PR for once this is merged.

**Comment by patrick96** [comment]
I'm not the biggest fan of iterating until no file is found. Is it guaranteed that all `thermal_zoneX` appear in order without gaps?

Also, the error message for when no zone is found is very confusing:

```
The file '/sys/class/thermal/thermal_zone1/type' does not exist
```

The error message should tell the user which zone type it failed to find (and maybe even list all available zone types).

---

> I have a few useful contributions which I can open PR for once this is merged.

Please open a [discussion topic](https://github.com/polybar/polybar/discussions/categories/ideas) for those. That way we can see how those will fit into polybar.

**Comment by xphoniex** [comment]
> I'm not the biggest fan of iterating until no file is found. Is it guaranteed that all `thermal_zoneX` appear in order without gaps?
> 
> Also, the error message for when no zone is found is very confusing:
> 
> ```
> The file '/sys/class/thermal/thermal_zone1/type' does not exist
> ```
> 
> The error message should tell the user which zone type it failed to find (and maybe even list all available zone types).
> 

There's [this](https://github.com/torvalds/linux/blob/268db333b561c77dee3feb6831806412293b4a7e/tools/lib/thermal/thermal.c#L34-L45) in kernel, but still not sure about gaps as there are ways to unregister sensors.

I suppose you've seen this issue before so I implemented what you suggested, check the [diff](https://github.com/polybar/polybar/compare/5b96d0f01dc08d9d434b6fc5baaa76e0d472b92e..6ced7cead38b9754ecd75c3628009a9aa9b2bf69) please.

**Comment by xphoniex** [comment]
reminder

File: src/modules/temperature.cpp
Comment (patrick96) [inline src/modules/temperature.cpp:52]:
```suggestion
        throw module_error("zone-type '" + m_zone_type +  "' was not found, available zone types: " + available_zones);
```

File: src/modules/temperature.cpp
Comment (patrick96) [inline src/modules/temperature.cpp:43]: Minor: Instead of manually constructing the string, we could store the types in a vector and finally construct the string using. `string_util::join(available_zones, ", ")`

File: src/modules/temperature.cpp
Comment (patrick96) [inline src/modules/temperature.cpp:35]: Optional, but could make this code cleaner: Instead of extracting the zone number, we could `file_util::glob` for `/sys/class/thermal/thermal_zone*` and just add `/type` for getting the type and `/temp` to get `m_path`, the zone number isn't really needed for the module itself.

Pick whatever solution you like better.

**Review by patrick96 — CHANGES_REQUESTED** [review]
Please also mention this feature in the [¢hangelog](https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md#changelog).

File: src/modules/temperature.cpp
Comment (xphoniex) [reply src/modules/temperature.cpp:35]: I kept the zone number as an effort to keep behavior consistent for when there are multiple sensors with the same name. We know we'll always pick the one with lowest number, which depending on hardware and kernel may or may not be random.

Do you still want me to change this?

Comment (patrick96) [reply src/modules/temperature.cpp:35]: Oh that's a good point. I think sorting the paths should be sufficient to keep consistency (even if it doesn't guarantee that the lowest number is picked).

It's up to you, whichever solution you like better

File: src/modules/temperature.cpp
Comment (xphoniex) [reply src/modules/temperature.cpp]: https://github.com/polybar/polybar/pull/2752/files#diff-4ecfc5580e380c6cc659596ee5bb12d8f8df8ceea72dae9559e1494503d54691R42

File: src/modules/temperature.cpp
Comment (xphoniex) [reply src/modules/temperature.cpp]: https://github.com/polybar/polybar/pull/2752/files#diff-4ecfc5580e380c6cc659596ee5bb12d8f8df8ceea72dae9559e1494503d54691R42

File: src/modules/temperature.cpp
Comment (xphoniex) [reply src/modules/temperature.cpp]: https://github.com/polybar/polybar/pull/2752/files#diff-4ecfc5580e380c6cc659596ee5bb12d8f8df8ceea72dae9559e1494503d54691R28

**Comment by xphoniex** [comment]
> Please also mention this feature in the [¢hangelog](https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md#changelog).

https://github.com/polybar/polybar/blob/0ba355760950b0b5932f46c82cac8858eb3b55a9/CHANGELOG.md

**Comment by patrick96** [comment]
Looks very good now. I'll not be at my computer for a few weeks, so I'll have to do the final review + testing once I get back.

**Comment by xphoniex** [comment]
bump to keep the PR in my **Recent activity**

**Review by patrick96 — APPROVED** [review]

**Comment by patrick96** [comment]
Thanks!

**Pull request merged**
